### ISSUE TITLE
Define a policy for error handling and reporting

### ISSUE THREAD

**Issue by pruzand** [ibm-js/sdk#1]
We need to decide how we properly handle and report errors in our delite/delitful code so that it is done in a consistent way throughout the lib.
Do we log every unexpected errors in a console.log ? Do we need a more "advanced" logging/tracing api ? 

Regarding how to handle errors, I guess it depends on the context where the error happens, which make difficult to write a policy, but there should be some common guidelines. For ex:
- do not catch silently exceptions.
- throw Error instances, (as opposed to simple object) so that they are properly handled by the browser
- ?

But what should we do with the following contexts:
- Should public methods systematically check the validity of their parameters (for ex, assert it's not null/undefined, etc) ? Or should we document what it expects and then...
- If a parameter has an incorrect value, should it return silently or throw error ?

To summarize, we need to define a common policy for error handling and reporting.

**Comment by wkeese** [comment]
To summarize from our phone meeting last night, IIUC the idea is:
- by default, do not do check properties or parameters for validity
- if there's a particular case where it's especially helpful to the user to check, then 
  - throw an exception rather than using console.log() or console.error()
  - add comment to the code about why you throwing the exception

An example of an exception is how delite/register() does checking that the base class is HTML*Element.   It's because otherwise the register() call will fail with a very cryptic exception, and the developer won't know how to fix the problem.

Patrick also thought that the List.scrollDirection was a case when an error should be thrown, since List (currently) only allows a subset of the values allowed by the ScrollableContainer mixin.  I didn't think that qualified since I thought the user could easily figure out that scrollDirection=horizontal doesn't (currently) work for List.

Patrick was going to write an official statement but I'm adding my notes before I forget.

**Comment by pruzand** [comment]
There's the case of the properties that accept only enumerated values : (e.g like a 'direction' property that would only accepts "left", "right" or "center").
if a property accepts a set of predefined values, then I would be in favor that it throws exception for incorrect values. (That's also why I prefer boolean parameter instead of a 2-values property, but that's another story).

**Comment by pruzand** [comment]
For the case above, if we don't want to use exception (I understand that's not that common in the JS world..) an alternative would be to fallback to a default value AND add a console.err message. This log would allow the dev to know there's something wrong, and could be stripped out at build time.

**Comment by wkeese** [comment]
You want to error check every enumerated property?   It sounds like too much code bloat to me.   In the meeting we agreed to only do error checking for exceptional cases.

**Comment by pruzand** [comment]
To be honest, I don't really have the answer. 
What I am sure is that I think we could do better than what it is done today in most of the JS framework (including dojo), which makes debugging faulty code very tedious, because of the js language nature.  I don't want to check systematically all the parameters of a method,  but there are some cases, as the one above, that could be a candidate.
Concretly, if the spec/doc clearly states what should be the value (for ex, "left/right/center") of a parameter, should the code accept any values and fallback to a default state if the value is wrong ? For me, this is similar to silently catch errors in the code and prevent the dev to fix his faulty code. 
But I know this is not something usual in the JS world.

**Comment by sbrunot** [comment]
I'm in favor of having at least two different errors processing rules, whatever the future rules that defines where to apply each of these processing rules:
1. If there is an error that is not recoverable (wrong value, the widget don't know what to do with it and it will result in the widget being in an "unknown" state), an Error should be throwned
2. If there is an error that is recoverable automatically (by switching to a default state / value), the widget should do so and log a warning (console.warn) to alert the user about it

I think that the overall aim of error handling is to help users to resolve problems when they occurs, but also to maximize the availability of application written using delite[ful].

**Comment by wkeese** [comment]
From 2014-4-8 meeting, we'll close this, integrating the decision from 2 months ago into https://github.com/ibm-js/sdk/blob/master/GUIDELINES.md.

**Comment by cjolif** [comment]
closed in 5236d5e75200aa3c615821b80a0028be5df45149

**Comment by AdrianVasiliu** [comment]
The guideline doesn't explicitly state the pattern to follow for error handling in the try/catch case. But by extrapolation from the rule about handling invalid method parameters which says an error should be thrown instead of using console.log/error, it would seem a catch block should rethrow the error as default policy (of course, if the error is unexpected). Now, the pain is that in some cases the rethrown error is silently catched by code outside ibm-js, such as dojo/Deferred (see the silent catch in its signalListener). In such cases I guess console.error() would be the way to go if we want to avoid a totally silent failure.

A way or another, I think it might be useful to add to the guideline an explicit point about error handling in try/catch.

**Comment by wkeese** [comment]
>  in some cases the rethrown error is silently catched by code outside ibm-js, such as dojo/Deferred (see the silent catch in its signalListener)

About that one, I assume/hope that everything will work correctly assuming that the code using the Deferred sets up a handler for when the Deferred is rejected?

**Comment by AdrianVasiliu** [comment]
Sure, and the policy could state that the unexpected rejection can/should provide error feedback to the user, just as in the general case of try-catching. And the policy could clarify the way feedback should be provided (rethrowing errors vs console.error etc.).
